src/Bonami/Collection/Monoid/Monoid.php

```diff
@@ -0,0 +1,31 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Bonami\Collection\Monoid;
+
+/** @phpstan-template-covariant A */
+interface Monoid
+{
+
+    /**
+     * A binary operation.
+     *
+     * It must follow monoid laws:
+     * - associativity - `concat(concat($a, $b), $c) === concat($a, concat($b, $c))` for any element `A`
+     * - identity law - `concat($a, getEmpty()) === concat(getEmpty(), $a) === $a`
+     *
+     * @phpstan-param A $a
+     * @phpstan-param A $b
+     *
+     * @phpstan-return A
+     */
+    public static function concat($a, $b);
+
+    /**
+     * Neutral element for binary concat operation
+     *
+     * @phpstan-return A
+     */
+    public static function getEmpty();
+}
```

Comment:
I'm still not sure, if I like that static method approach, maybe I prefer instance methods + good old static `getInstance` ... have to think about it for a while.

Reply:
Monoids use to be equipped w/ contramap function / method, which allows to transform `Monoid<A>` to `Monoid<B>` with the help of `B => A` function


I wonder, how to implement it when methods are static.. If it were instance methods, I'd do:
```php
/**
 * @template B
 * @param callable(B): A $mapper
 * @return Monoid<B>
 */
public function contramap($mapper) {
  return new class($this, $mapper) extends Monoid() {
    private Monoid $class;
    private callable $mapper;
    public function __construct(Monoid $class, callback $mapper) {
      $this->class = $class;
      $this->mapper = $mapper;
    }
    public function concat($a, $b) { return $this->class->concat(($this->mapper)($a), ($this->mapper)($b)); }
    public function getEmpty() { return $this->class->getEmpty(); }
  } 
}
...
```

Then you can do something like
```php
$userOrderCountMonoid = $intMonoid->contramap(fn(User $u): int => $u->getOrderCount());
```
(not the best example, I know ;-))